Add `get_magnetic_spacegroup_type_from_symmetry` and introduce dataclass output in python interface (#485)
spglib/spglib: ChangeLog.md

## \[Unreleased\]

### Python API

- Add `spglib.get_magnetic_spacegroup_type_from_symmetry` for getting a magnetic space-group type information from magnetic symmetry operations
- `spglib.get_spacegroup_type` and `spglib.get_spacegroup_type_from_symmetry` now return a `spglib.SpaceGroupType` object, inheriting `dataclasses.dataclass`, instead of a plain `dict`. This object also has the `dict`-like interface, for example

```python
spgtype: SpaceGroupType | None = get_spacegroup_type(cell)
assert spgtype is not None
print(spgtype['number'])
print(list(spgtype.items()))
```

> However, we recommend using the `dataclass` interface:

Reviewer: This would be better in a `### Main changes` (particularly if there are deprecation notices). In the `Python API`, just mention which methods were changed.

## \[Unreleased\]

### Main changes

We now provide proper class types with type-hints (using `dataclasses.dataclass`) for the output data structure.
See the Python API section for details for the list of new classes and affected functions.

For backward compatibility, the `dict`-like interface is still available, but it is planned to be removed:

```python
spgtype: SpaceGroupType | None = get_spacegroup_type(cell)
assert spgtype is not None
print(spgtype['number'])
print(list(spgtype.items()))
```

However, we recommend using the `dataclass` interface:

```python
spgtype: SpaceGroupType | None = get_spacegroup_type(cell)
assert spgtype is not None
print(spgtype.number)

# If you really need dictionary, ...
from dataclasses import asdict
print(asdict(spgtype))
```

### Python API

- Add `spglib.get_magnetic_spacegroup_type_from_symmetry` for getting a magnetic space-group type information from magnetic symmetry operations

> - `spglib.get_spacegroup_type` and `spglib.get_spacegroup_type_from_symmetry` now return a `spglib.SpaceGroupType` object.

Reviewer: ```suggestion
- Added classes: `SpaceGroupType` and `MagneticSpaceGroupType`
- Added function: `get_magnetic_spacegroup_type_from_symmetry`
- Changed functions: `get_spacegroup_type`, `get_spacegroup_type_from_symmetry`, `get_magnetic_spacegroup_type`,
  and `get_magnetic_spacegroup_type_from_symmetry` now return the equivalent
  `SpaceGroupType` or `MagneticSpaceGroupType` object
```

## \[Unreleased\]

### Main changes

> In Python API, `spglib.get_spacegroup_type` and `spglib.get_spacegroup_type_from_symmetry` now return a `spglib.SpaceGroupType` object, inheriting `dataclasses.dataclass`, instead of a plain `dict`.
> The `dict`-like interface is deprecated and will be removed in the future.

Reviewer: ```suggestion
We now provide proper class types with type-hints (using `dataclasses.dataclass`) for the output data structure.
See the Python API section for details for the list of new classes and affected functions. 
```

Also a bit of re-arranging on the Python API section.

> For backward compatibility, the `dict`-like interface is still available:

Reviewer: ```suggestion
For backward compatibility, the `dict`-like interface is still available, but it is planned to be removed:
```

```python
spgtype: SpaceGroupType | None = get_spacegroup_type(cell)
assert spgtype is not None
print(spgtype['number'])
print(list(spgtype.items()))
```

However, we recommend using the `dataclass` interface:

```python
spgtype: SpaceGroupType | None = get_spacegroup_type(cell)
assert spgtype is not None
print(spgtype.number)

# If you really need dictionary, ...
from dataclasses import asdict
print(asdict(spgtype))
```

> We consider migrating plain `dict` into `dataclasses.dataclass` for the returned objects of other Python APIs.

Reviewer: I think we can address this within the `2.5.0` release

Author: OK. I'll remove this line.
